txpipe/calibrate.py

```diff
@@ -0,0 +1,151 @@
+        subtract_mean_shear = self.config["subtract_mean_shear"]
+
+        # Prepare the output file, and
+        output_file, splitter, nbin = self.setup_output()
+
+        #  Load the calibrators.  If using the true shear no calibration
+        # is needed
+        tomo_file = self.get_input("shear_tomography_catalog")
+        cals, cal2d = Calibrator.load(tomo_file, null=use_true)
+
+        # These are always named the same
+        cat_cols = ["ra", "dec", "weight"]
+        # The catalog columns are named differently in different cases
+        #  Get the correct shear catalogs
+        if use_true:
+            cat_cols += ["true_g1", "true_g2"]
+        elif cat_type == "metacal":
+            cat_cols += ["mcal_g1", "mcal_g2"]
+        else:
+            cat_cols += ["g1", "g2"]
+
+        output_cols = ["ra", "dec", "g1", "g2", "weight"]
+
+        # We parallelize by bin.  This isn't ideal but we don't know the number
+        # of objects in each bin per chunk, so we can't parallelize in full.  This
+        #  is a quick stage though.
+        bins = list(range(nbin)) + ["all"]
+        my_bins = list(self.split_tasks_by_rank(bins))
+
+        # Print out which bins this proc will do, also as a prompt to the user
+        #  in case they're wondering why adding procs doesn't help
+        if my_bins:
+            my_bins_text = ", ".join(str(x) for x in my_bins)
+            print(f"Process {self.rank} collating bins: [{my_bins_text}]")
+        else:
+            print(f"Note: Process {self.rank} will not do anything.")
+
+        # make the iterator that loops through data
+        it = self.combined_iterators(
+            self.config["chunk_rows"],
+            # first file
+            "shear_catalog",
+            "shear",
+            cat_cols,
+            "shear_tomography_catalog",
+            "tomography",
+            ["source_bin"],
+            parallel=False,
+        )
+
+        #  Main loop
+        for s, e, data in it:
+
+            if self.rank == 0:
+                print(f"Rank 0 processing data {s:,} - {e:,}")
+            # Rename mcal_g1 -> g1 etc
+            self.rename_metacal(data)
+
+            #  Now output the calibrated bin data for this processor
+            for b in my_bins:
+
+                # Select objects to go in this bin
+                if b == "all":
+                    # the 2D case is any object from any other bin
+                    w = np.where(data["source_bin"] >= 0)
+                    cal = cal2d
+                else:
+                    # otherwise just objects in this bin
+                    w = np.where(data["source_bin"] == b)
+                    cal = cals[b]
+
+                # Cut down the data to just this selection for output
+                d = {name: data[name][w] for name in output_cols}
+
+                # Calibrate the shear columns
+                d["g1"], d["g2"] = cal.apply(d["g1"], d["g2"], subtract_mean=subtract_mean_shear)
```

Comment:
Do I understand this correctly that the individual objects will be saved with object-based calibrated shear? a couple things...  1) do we ever want to save the calibration factors for debugging purpose? 2) I thought the way metacal functions, the calibration needs to be estimated via an ensemble of galaxies (either at the 1pt or 2pt level, depending on the statistics) -- is this still consistent with that?